development/image-syncer/main.go

```diff
@@ -0,0 +1,228 @@
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/jamiealquiza/envy"
+	parser "github.com/novln/docker-parser"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	log = logrus.New()
+)
+
+type SyncDef struct {
+	TargetRepoPrefix string `yaml:"targetRepoPrefix"`
+	Images           []Image
+}
+
+type Image struct {
+	Source string
+}
+
+type Config struct {
+	ImagesFile    string
+	TargetKeyFile string
+	DryRun        bool
+}
+
+func getAuthString(user, password string) (string, error) {
+	authConfig := types.AuthConfig{
+		Username: user,
+		Password: password,
+	}
+	encodedJSON, err := json.Marshal(authConfig)
+	if err != nil {
+		return "", err
+	}
+	return base64.URLEncoding.EncodeToString(encodedJSON), nil
+}
+
+func getImageIdAndRepoDigest(ctx context.Context, cli *client.Client, authString, image string) (string, string, error) {
+	if reader, err := cli.ImagePull(ctx, image, types.ImagePullOptions{}); err != nil {
+		return "", "", fmt.Errorf("image pull failed: %w", err)
+	} else {
+		w := log.WriterLevel(logrus.DebugLevel)
+		defer w.Close()
+		io.Copy(w, reader)
+	}
+
+	details, _, err := cli.ImageInspectWithRaw(ctx, image)
+	if err != nil {
+		return "", "", fmt.Errorf("image inspection failed: %w", err)
+	}
+
+	ref, err := parser.Parse(image)
+	imageRepository := ref.Repository()
+
+	for _, url := range details.RepoDigests {
+		ref, err := parser.Parse(url)
+		if err != nil {
+			return "", "", fmt.Errorf("url parsing failed: %w", err)
+		}
+		if imageRepository == ref.Repository() {
+			return details.ID, ref.Tag(), nil
+		}
+	}
+
+	return "", "", errors.New(fmt.Sprintf("Unable to find digest for '%s'", image))
+}
+
+func safeCopyImage(ctx context.Context, cli *client.Client, authString, source, target string, dryRun bool) error {
+	if source == "" {
+		return errors.New("source image can not be empty")
+	}
+	log.Infof("Source image: %s", source)
+	sourceId, sourceDigest, err := getImageIdAndRepoDigest(ctx, cli, authString, source)
+	if err != nil {
+		return err
+	}
+	log.Infof("Source ID: %s", sourceId)
+	log.Infof("Source repo digest: %s", sourceDigest)
+
+	log.Infof("Target image: %s", target)
+	targetId, targetDigest, err := getImageIdAndRepoDigest(ctx, cli, authString, target)
+	if isImageNotFoundError(err) {
+		log.Info("Target image does not exist")
+		if dryRun {
+			log.Info("Dry-run mode - tagging and pushing skipped")
+			return nil
+		}
+		if err = cli.ImageTag(ctx, source, target); err != nil {
+			return err
+		}
+		log.Info("Image re-tagged")
+		log.Info("Pushing to target repo")
+		if reader, err := cli.ImagePush(ctx, target, types.ImagePushOptions{RegistryAuth: authString}); err != nil {
```

Comment:
For logging in and tracking errors if the login credentials are wrong can we use for example [RegistryLogin method](https://godoc.org/github.com/docker/docker/client#Client.RegistryLogin) of Docker Client? If the credentials are wrong it returns `unauthorizedError` when the authorisation fails.

Reply:
good point, I will test this method against malformed json file